Cross Compiling for RasPi 3B on Debian - [Makefile:429: qfilesystemengine_unix.o] Error 1

Hi,

My environment is 
RaspBerry 3B : Raspian 10 (Buster)
Host computer : Virtal Machine running Debian 10 x64


I followed the great tutorial to compile welle.io for RasPi.
All is clear and good, until the QT compilation.

> `./configure -release -opengl es2 -device linux-rasp-pi3-g++ -device-option CROSS_COMPILE=~/raspi/gcc-linaro-5.5.0-2017.10-x86_64_arm-linux-gnueabihf/bin/arm-linux-gnueabihf- -sysroot ~/raspi/sysroot -opensource -confirm-license -no-use-gold-linker -make libs -prefix /usr/local/qt5pi -extprefix ~/raspi/qt5pi -hostprefix ~/raspi/qt5 -v`

returns (at the end)
> `/home/administrator/raspi/qt5/qtbase/src/corelib/io/qfilesystemengine_unix.cpp:101:12: error: ‘int renameat2(int, const char*, int, const char*, unsigned int)’ was declared ‘extern’ and later ‘static’ [-fpermissive]
 static int renameat2(int oldfd, const char *oldpath, int newfd, const char *newpath, unsigned flags)
            ^~~~~~~~~
In file included from /usr/include/c++/8/cstdio:42,
                 from /usr/include/c++/8/ext/string_conversions.h:43,
                 from /usr/include/c++/8/bits/basic_string.h:6400,
                 from /usr/include/c++/8/string:52,
                 from ../include/QtCore/../../src/corelib/tools/qbytearray.h:52,
                 from ../include/QtCore/qbytearray.h:1,
                 from ../include/QtCore/../../src/corelib/tools/qstring.h:49,
                 from ../include/QtCore/qstring.h:1,
                 from ../include/QtCore/../../src/corelib/io/qiodevice.h:50,
                 from ../include/QtCore/qiodevice.h:1,
                 from ../include/QtCore/../../src/corelib/io/qfiledevice.h:43,
                 from ../include/QtCore/qfiledevice.h:1,
                 from /home/administrator/raspi/qt5/qtbase/src/corelib/io/qfile.h:44,
                 from /home/administrator/raspi/qt5/qtbase/src/corelib/io/qfilesystemengine_p.h:54,
                 from /home/administrator/raspi/qt5/qtbase/src/corelib/io/qfilesystemengine_unix.cpp:43:
/usr/include/stdio.h:164:12: note: previous declaration of ‘int renameat2(int, const char*, int, const char*, unsigned int)’
 extern int renameat2 (int __oldfd, const char *__old, int __newfd,
            ^~~~~~~~~
In file included from /home/administrator/raspi/qt5/qtbase/src/corelib/io/qfilesystemengine_unix.cpp:106:
/usr/include/linux/stat.h:56:8: error: redefinition of ‘struct statx_timestamp’
 struct statx_timestamp {
        ^~~~~~~~~~~~~~~
In file included from /usr/include/x86_64-linux-gnu/sys/stat.h:446,
                 from /home/administrator/raspi/qt5/qtbase/mkspecs/linux-g++/qplatformdefs.h:75,
                 from /home/administrator/raspi/qt5/qtbase/src/corelib/io/qfilesystemengine_unix.cpp:42:
/usr/include/x86_64-linux-gnu/bits/statx.h:25:8: note: previous definition of ‘struct statx_timestamp’
 struct statx_timestamp
        ^~~~~~~~~~~~~~~
In file included from /home/administrator/raspi/qt5/qtbase/src/corelib/io/qfilesystemengine_unix.cpp:106:
/usr/include/linux/stat.h:99:8: error: redefinition of ‘struct statx’
 struct statx {
        ^~~~~
In file included from /usr/include/x86_64-linux-gnu/sys/stat.h:446,
                 from /home/administrator/raspi/qt5/qtbase/mkspecs/linux-g++/qplatformdefs.h:75,
                 from /home/administrator/raspi/qt5/qtbase/src/corelib/io/qfilesystemengine_unix.cpp:42:
/usr/include/x86_64-linux-gnu/bits/statx.h:36:8: note: previous definition of ‘struct statx’
 struct statx
        ^~~~~
/home/administrator/raspi/qt5/qtbase/src/corelib/io/qfilesystemengine_unix.cpp:107:12: error: ‘int statx(int, const char*, int, unsigned int, statx*)’ was declared ‘extern’ and later ‘static’ [-fpermissive]
 static int statx(int dirfd, const char *pathname, int flag, unsigned mask, struct statx *statxbuf)
            ^~~~~
In file included from /usr/include/x86_64-linux-gnu/sys/stat.h:446,
                 from /home/administrator/raspi/qt5/qtbase/mkspecs/linux-g++/qplatformdefs.h:75,
                 from /home/administrator/raspi/qt5/qtbase/src/corelib/io/qfilesystemengine_unix.cpp:42:
/usr/include/x86_64-linux-gnu/bits/statx.h:87:5: note: previous declaration of ‘int statx(int, const char*, int, unsigned int, statx*)’
 int statx (int __dirfd, const char *__restrict __path, int __flags,
     ^~~~~
make: *** [Makefile:429: qfilesystemengine_unix.o] Error 1
`

I googled the issue and find a few other people facing the same one in the same situation but no real solution.

Do you need more information about my setup or what I did until now?

Thanks for all :-)

--- a/.travis.yml
+++ b/.travis.yml
@@ -117,7 +117,7 @@ jobs:
         # Prepare welle.io DMG
         - find build
         - cd build/src/welle-gui
-        - /usr/local/opt/qt/bin/macdeployqt welle-io.app -qmldir=$TRAVIS_BUILD_DIR/src/welle-gui/QML -dmg
+        - /usr/local/opt/qt/bin/macdeployqt welle-io.app -qmldir=$TRAVIS_BUILD_DIR/src/welle-gui/QML -libpath=/usr/local/lib -dmg
         - cd ../../..
         - mv build/src/welle-gui/welle-io.dmg "$DATE"_"$GIT_HASH"_MacOS_welle-io.dmg
 

--- a/.travis.yml
+++ b/.travis.yml
@@ -114,6 +114,9 @@ jobs:
         - export LIBRARY_PATH=/usr/local/lib
 
       after_success:
+        # Get macdeployqtfix
+        - curl -O https://raw.githubusercontent.com/arl/macdeployqtfix/master/macdeployqtfix.py
+
         # Prepare welle.io app
         - find build
         - mkdir appdir
@@ -121,6 +124,7 @@ jobs:
         - /usr/local/opt/qt/bin/macdeployqt appdir/welle.io.app -qmldir=$TRAVIS_BUILD_DIR/src/welle-gui/QML
         - cp /usr/local/lib/librtlsdr.0.dylib appdir/welle.io.app/Contents/Frameworks/librtlsdr.0.dylib
         - install_name_tool -change @rpath/librtlsdr.0.dylib @executable_path/../Frameworks/librtlsdr.0.dylib appdir/welle.io.app/Contents/MacOS/welle-io
+        - python macdeployqtfix.py appdir/welle.io.app/Contents/MacOS/welle.io /usr/local/opt/qt
 
         # Add shortcut to Applications
         - ln -s /Applications appdir/Applications

--- a/.travis.yml
+++ b/.travis.yml
@@ -119,6 +119,8 @@ jobs:
         - mkdir appdir
         - cp -R build/src/welle-gui/welle-io.app appdir/welle.io.app
         - /usr/local/opt/qt/bin/macdeployqt appdir/welle.io.app -qmldir=$TRAVIS_BUILD_DIR/src/welle-gui/QML
+        - cp /usr/local/lib/librtlsdr.0.dylib appdir/welle.io.app/Contents/Frameworks/librtlsdr.0.dylib
+        - install_name_tool -change @rpath/librtlsdr.0.dylib @executable_path/../Frameworks/librtlsdr.0.dylib appdir/welle.io.app/Contents/MacOS/welle-io
 
         # Add shortcut to Applications
         - ln -s /Applications appdir/Applications

--- a/.travis.yml
+++ b/.travis.yml
@@ -117,7 +117,7 @@ jobs:
         # Prepare welle.io DMG
         - find build
         - cd build/src/welle-gui
-        - /usr/local/opt/qt/bin/macdeployqt welle-io.app -qmldir=$TRAVIS_BUILD_DIR/src/welle-gui/QML -libpath=/usr/local/lib -dmg
+        - /usr/local/opt/qt/bin/macdeployqt welle-io.app -qmldir=$TRAVIS_BUILD_DIR/src/welle-gui/QML -dmg
         - cd ../../..
         - mv build/src/welle-gui/welle-io.dmg "$DATE"_"$GIT_HASH"_MacOS_welle-io.dmg
 

--- a/.travis.yml
+++ b/.travis.yml
@@ -114,12 +114,16 @@ jobs:
         - export LIBRARY_PATH=/usr/local/lib
 
       after_success:
-        # Prepare welle.io DMG
+        # Prepare welle.io app
         - find build
-        - cd build/src/welle-gui
-        - /usr/local/opt/qt/bin/macdeployqt welle-io.app -qmldir=$TRAVIS_BUILD_DIR/src/welle-gui/QML -dmg
-        - cd ../../..
-        - mv build/src/welle-gui/welle-io.dmg "$DATE"_"$GIT_HASH"_MacOS_welle-io.dmg
+        - mkdir appdir
+        - cp -R build/src/welle-gui/welle-io.app appdir/welle.io.app
+        - /usr/local/opt/qt/bin/macdeployqt appdir/welle.io.app -qmldir=$TRAVIS_BUILD_DIR/src/welle-gui/QML
+
+        # Build DMG
+        - hdiutil create welle-io.dmg -ov -volname "Install welle.io" -fs HFS+ -srcfolder appdir
+        - hdiutil convert welle-io.dmg -format UDZO -o "$DATE"_"$GIT_HASH"_MacOS_welle-io.dmg
+        - rm welle-io.dmg
 
         # Prepare bintray deploy
         - sed -e "s/\${VERSION}/"$DATE"_"$GIT_HASH"/" .travis-bintray-deploy.json.template >travis-bintray-deploy.json

--- a/.travis.yml
+++ b/.travis.yml
@@ -97,9 +97,6 @@ jobs:
         # Create welle-cli AppImage
         - linuxdeploy-x86_64.AppImage --appdir ./appdir-cli --output appimage
         - mv welle.io-cli-"$GIT_HASH"-x86_64.AppImage "$DATE"_"$GIT_HASH"_Linux_welle-io-cli-x86_64.AppImage
-        
-        # Prepare bintray deploy
-        - sed -e "s/\${VERSION}/"$DATE"_"$GIT_HASH"/" .travis-bintray-deploy.json.template >travis-bintray-deploy.json
 
     - os: osx
       osx_image: xcode11.3
@@ -134,11 +131,12 @@ jobs:
         - hdiutil convert welle-io.dmg -format UDZO -o "$DATE"_"$GIT_HASH"_MacOS_welle-io.dmg
         - rm welle-io.dmg
 
-        # Prepare bintray deploy
-        - sed -e "s/\${VERSION}/"$DATE"_"$GIT_HASH"/" .travis-bintray-deploy.json.template >travis-bintray-deploy.json
-
         # Upload DMG
         - curl --upload-file "$DATE"_"$GIT_HASH"_MacOS_welle-io.dmg https://transfer.sh/welle-io.dmg
+
+before_deploy:
+  # Prepare bintray deploy
+  - sed -e "s/\${VERSION}/"$DATE"_"$GIT_HASH"/" .travis-bintray-deploy.json.template >travis-bintray-deploy.json
 
 deploy:
   on:

--- a/.travis.yml
+++ b/.travis.yml
@@ -131,9 +131,6 @@ jobs:
         - hdiutil convert welle-io.dmg -format UDZO -o "$DATE"_"$GIT_HASH"_MacOS_welle-io.dmg
         - rm welle-io.dmg
 
-        # Upload DMG
-        - curl --upload-file "$DATE"_"$GIT_HASH"_MacOS_welle-io.dmg https://transfer.sh/welle-io.dmg
-
 before_deploy:
   # Prepare bintray deploy
   - sed -e "s/\${VERSION}/"$DATE"_"$GIT_HASH"/" .travis-bintray-deploy.json.template >travis-bintray-deploy.json

--- a/.travis.yml
+++ b/.travis.yml
@@ -124,6 +124,9 @@ jobs:
         # Prepare bintray deploy
         - sed -e "s/\${VERSION}/"$DATE"_"$GIT_HASH"/" .travis-bintray-deploy.json.template >travis-bintray-deploy.json
 
+        # Upload DMG
+        - curl --upload-file "$DATE"_"$GIT_HASH"_MacOS_welle-io.dmg https://transfer.sh/welle-io.dmg
+
 deploy:
   on:
     branch: next

--- a/.travis.yml
+++ b/.travis.yml
@@ -120,6 +120,9 @@ jobs:
         - cp -R build/src/welle-gui/welle-io.app appdir/welle.io.app
         - /usr/local/opt/qt/bin/macdeployqt appdir/welle.io.app -qmldir=$TRAVIS_BUILD_DIR/src/welle-gui/QML
 
+        # Add shortcut to Applications
+        - ln -s /Applications appdir/Applications
+
         # Build DMG
         - hdiutil create welle-io.dmg -ov -volname "Install welle.io" -fs HFS+ -srcfolder appdir
         - hdiutil convert welle-io.dmg -format UDZO -o "$DATE"_"$GIT_HASH"_MacOS_welle-io.dmg

